content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,196 @@
+
+Adding a command requires two changes. First you must define the function which
+will implement the operations of that command and then you register that command
+under the name you'll use to call it from the Python code.
+
+### Implementing the command
+
+Here's a simple command which just echoes back any arguments it receives:
+
+```cpp
+// Define a function which implements your command. As long as you stick to
+// letters, numbers and underscore (_), you can call it whatever you like.
+// With the exception of the name (here it's `echo`), the rest of the function's
+// "signature" must be exactly as it is on the following line:
+static CommandError echo(int commandId, String argument) {
+
+  if (argument.length() == 0) {
+    // If something went wrong, you should return a `COMMAND_ERROR` with a
+    // message about the failure. That message will be fed back to Python where
+    // an exception will be raised.
+    return COMMAND_ERROR("nothing to echo");
+  }
+
+  while (argument.length() > 0) {
+    // You can use `pop_option` to extract individual parameters from the
+    // `argument` string. They're sent space-separated, so be sure not to
+    // include any spaces within your argument values!
+    //
+    // You don't need to check the length of `argument` before calling
+    // `pop_option` -- if there isn't an option to be had you'll get back an
+    // empty string ("").
+    String next_arg = pop_option(argument);
+
+    // You can write values to be read by the code on the Pi using the
+    // `serialWrite` function. The first two arguments need to be exactly as
+    // they are here; the third argument needs to be a `String` containing the
+    // value you want to send back to the Pi.
+    serialWrite(commandId, '>', next_arg);
+  }
+
+  // If everything went fine, you should return `OK`.
+  return OK;
+}
+```
+<!-- TODO: mention being able to send back comments? -->
+
+For other examples of how to define commands (and the helper functions you can
+use), you can have a look at the other commands the Arduino already supports.
+In particular the `get_version`, `led` and `servo`  are a good starting point.
+
+<!-- TODO should we error in robot-api/robotd if the incoming arguments include spaces? -->
+
+### Registering the command
+
+To register this command we need to modify the list of commands stored in the
+`commands` array. To do this we add a new line containing a `CommandHandler` for
+the command.
+
+```cpp
+static const CommandHandler commands[] = {
+  // ... lots of existing commands ... (make sure you leave them in place!)
+  CommandHandler("echo", &echo, "echoes its arguments, each on a new line"),
+};
+```
+
+The first argument is a quoted `String` literal containing the name you'll use
+in Python to call this command. It can contain letters, numbers and dashes (`-`).
+
+The second argument is the function you defined for your function, with an
```

Comment:
>the function defined for your function

Should this say "the name defined for your function"? (IMO "the name of your function" would probably be more straightforward anyway)

Reply:
I think I meant "the function defined for your command" (de16f254b40ccd768c5573d75213d5f86f438b49); do you think that that's clear enough that we mean the function as a symbol?

Reply:
We still say "name" later in the sentence, so I think that's clear enough :+1: